lib/cousins.ml

```diff
@@ -246,40 +292,65 @@ let init_cousins_cnt base max_a_l max_d_l p =
       in
       cousins_t := Some t';
       cousins_dates_t := Some d_t';
+      flush stderr;
       (t', d_t')
 
+(* for cousins_dates.(l1).(l2) determine min or max date *)
+let min_max_date conf base p min_max l1 l2 =
+  let _cousins_cnt, cousins_dates =
+    match (!cousins_t, !cousins_dates_t) with
+    | Some t, Some d_t -> (t, d_t)
+    | _, _ -> init_cousins_cnt conf base p
+  in
+  let i = try int_of_string l1 with Failure _ -> -1 in
+  let j = try int_of_string l2 with Failure _ -> -1 in
+  match (i, j) with
+  | -1, _ | _, -1 -> None
+  | _, _ ->
+      let min, max =
+        if
+          i + 1 > Array.length cousins_dates
+          || j + 1 > Array.length cousins_dates.(i)
+        then (-1, -1)
+        else cousins_dates.(i).(j)
+      in
+      if min_max then Some min else Some max
+
 (* determine non empty max ancestor level (max_i)
    and non empty max descendant level
 *)
-let max_l1_l2 base max_a_l max_d_l p =
+let max_l1_l2 conf base p =
   let cousins_cnt, _cousins_dates =
     match (!cousins_t, !cousins_dates_t) with
     | Some t, Some d_t -> (t, d_t)
-    | _, _ -> init_cousins_cnt base max_a_l max_d_l p
+    | _, _ -> init_cousins_cnt conf base p
   in
-  let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
-  let max_i =
+  let max_i = Array.length cousins_cnt - 1 in
+  let max_j = Array.length cousins_cnt.(0) - 1 in
+  let max_a =
     let rec loop0 i =
-      if cousins_cnt.(i).(0) <> [] && i < max_l1 then loop0 (i + 1) else i
+      if cousins_cnt.(i).(0) <> [] && i < max_i - 1 then loop0 (i + 1) else i
     in
     loop0 0
   in
   let rec loop i j =
     if cousins_cnt.(i).(j) <> [] then
-      if j < max_l2 then loop i (j + 1) else (max_i, j - i)
-    else if i < max_l1 && j < max_l2 then loop (i + 1) (j + 1)
-    else (max_i, j - i)
+      if j < max_j then loop i (j + 1) else (max_a, j - i)
+    else if i < max_i && j < max_j then loop (i + 1) (j + 1)
+    else (max_a, j - i)
   in
   loop 0 0
 
-let cousins_l1_l2_aux base max_a_l max_d_l l1 l2 p =
+let cousins_l1_l2_aux conf base l1 l2 p =
   let il1 = int_of_string l1 in
   let il2 = int_of_string l2 in
-  let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
-  if il1 <= max_l1 && il2 - il1 <= max_l2 then
-    let cousins_cnt, _cousins_dates = init_cousins_cnt base max_a_l max_d_l p in
-    (* gros calcul *)
-    Some cousins_cnt.(il1).(il2)
+  let cousins_cnt, _cousins_dates =
+    match (!cousins_t, !cousins_dates_t) with
+    | Some t, Some d_t -> (t, d_t)
+    | _, _ -> init_cousins_cnt conf base p
+  in
+  if il1 < Array.length cousins_cnt && il2 - il1 < Array.length cousins_cnt.(0)
+  then Some cousins_cnt.(il1).(il2)
   else None
 
 (* create a new l (ip, (ifamll, iancl, cnt), lev) from (ip, ifaml, ianc, lev) *)
```

Comment:
Maybe a record type instead of a big tuple will make the code easier to understand and work with.
This code will probably benefit from more abstraction

Reply:
can you help?

Reply:
I wish but I don't have a good understanding of the code in Cousins

What you should do is start by looking at the types, in cousins.mli, you can see long and not trivial to understand type:
```ocaml
(Gwdb_driver.iper * Gwdb_driver.ifam list * Gwdb_driver.iper * int list) list
```
that appear multiple times

you can start to give a name to this type
in cousins.ml and  in cousins.mli  :
```ocaml
type machin_truc_cousin = (Gwdb_driver.iper * Gwdb_driver.ifam list * Gwdb_driver.iper * int list) list
```
and replace occurrence of this type in the mli, this will probably make things clearer for someone not aware of all the details and the implementations of the module

after that if there are other types like this that are complex and appears multiple times you can try to do the same for them too

just giving names to types can help a lot for other reading the code!

then,
after doing that, since tuple can be a little confusing because they do not have a name for each variable inside, you can use a record type instead of a tuple and choose nice descriptive names:
```ocaml
type machin_record = {iper_1: Gwdb_driver.iper;
                                     family_list: Gwdb_driver.ifam list;
                                     iper_2: Gwdb_driver.iper;
                                     liste_entiers: int list}
type machin_truc_cousin = machin_record list
```